### PR TITLE
Technigo Boot Camp week 19: Happy Thoughts API - Ida Sylvén

### PR DESCRIPTION
**Pull request by idasylven**
Link to deployed app: https://is-happy-thoughts.herokuapp.com/
Link to fullstack project: https://pensive-shirley-fd39b0.netlify.app/

My reflections on how this week's project turned out: I thought this was a fun assignment, specially since it was relating to the previous Happy Thoughts project. It was nice to feel a bit more comfortable with the syntax in general and with the process of deploying the API and the database itself.  

Things I'd like to have clarified or explained in more detail: Nothing in particular at this current stage.

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -1,30 +1,138 @@
 import express from 'express'
-import bodyParser from 'body-parser'
 import cors from 'cors'
 import mongoose from 'mongoose'
+import dotenv from 'dotenv'
+import listEndpoints from 'express-list-endpoints'
```
Comment (TobiasHagberg) [inline server.js:5]: Always a nice touch for simple documentation 📑

File: server.js [hunk 2]
```diff
@@ -1,30 +1,138 @@
...
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
+// Defining the port the app will run on
 const port = process.env.PORT || 8080
 const app = express()
 
-// Add middlewares to enable cors and json body parsing
+// Middlewares to enable cors and json body parsing
 app.use(cors())
```
Comment (TobiasHagberg) [inline server.js:18]: Quick reminder from this weeks labs. The use of cors without specifying exceptions is what enables access from all domains.

File: server.js [hunk 3]
```diff
@@ -1,30 +1,138 @@
...
-app.use(bodyParser.json())
+app.use(express.json())
+
+const thoughtSchema = new mongoose.Schema({
+  message: {
+    type: String,
+    required: [true, "A message is required, please type a happy thought :)"],
+    minlenght: [5, "Sorry, your message needs to have at least 5 characters"],
+    maxlenght: [140, "Sorry, you message is unfortunately exceeding the limit of 140 characters"]
```
Comment (TobiasHagberg) [inline server.js:26]: Like the extra effort put into these explanations

File: .env [hunk 4]
```diff
@@ -0,0 +1,3 @@
+MONGO_URL = 'mongodb+srv://noveltyUser:bwb6FhIxLTk7trUx@cluster0.oajux.mongodb.net/is-happy-thoughts?retryWrites=true&w=majority'
+
+RESET_DB = true
```
Comment (TobiasHagberg) [inline .env:3]: You have forgot to list your .env to the gitnore file and are therefore sharing sensitive data with the world 🌎😬

**Review by TobiasHagberg — APPROVED** [review]
A nice project which meets all the requirements for the week. It's especially fun to read about the added motivation of being a full stack developer. Well done 👏😊🎉

File: server.js (hunk 3 above)
Comment (KarolCislowski) [inline server.js:26]: Nicely configured field 👍🏻

File: server.js [hunk 5]
```diff
@@ -1,30 +1,138 @@
...
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send(listEndpoints(app))
 })
 
-// Start the server
+// Endpoint to get the most recent 20 thoughts in descending order
+app.get('/thoughts', async (req, res) => {
+  try {
+    const allThoughts = await Thought.find().sort({ createdAt: -1 }).limit(20)
```
Comment (KarolCislowski) [inline server.js:48]: 👍🏻

File: server.js [hunk 6]
```diff
@@ -1,30 +1,138 @@
...
+    res.status(400).json({ message: 'Invalid request', error })
+  }
+})
+
+// Endpoint to update likes/hearts 
+app.post('/thoughts/:thoughtId/like', async (req, res) => {
+  const { thoughtId } = req.params
+
+  try {
+    const updatedThought = await Thought.findOneAndUpdate(thoughtId, { $inc: { hearts: 1 } }, { new: true })
```
Comment (KarolCislowski) [inline server.js:78]: 💯

File: server.js [hunk 7]
```diff
@@ -1,30 +1,138 @@
...
+//     const updatedThought = await Thought.findByIdAndUpdate(id, req.body, { new: true })
+//     if (updatedThought) {
+//       res.json(updatedThought)
+//     } else {
+//       res.status(404).json({ message: 'Not found' })
+//     }
+//   } catch (error) {
+//     res.status(400).json({ message: 'Invalid request', error })
+//   }
+// })
```
Comment (KarolCislowski) [inline server.js:138]: Remember to restrict access to these endpoints because not all users should have access to these functionalities 🤓

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
The code looks great but you need to provide a link to the frontend connected to it 😸

**Comment by idasylven** [comment]
Thank you for your comments! 💫  I didn't take connecting them as a necessity at the point of submission, as it wasn't listed among the general or blue level requirements. But I did indeed have future plans to do so - and now I have! Please find updated README and Pull Request with the link of the connected frontend project 🔄  👍🏻  😊

**Review by KarolCislowski — APPROVED** [review]
Great APP

**Pull request closed without merging**